ursine/header.py

```diff
@@ -0,0 +1,106 @@
+import copy
+import random
+import string
+from .header_parsing import parse_header
+
+
+def random_tag():
+    return ''.join([random.choice(string.hexdigits) for _ in range(16)])
+
+
+class HeaderError(Exception):
+    pass
+
+
+class Header:
+    __slots__ = (
+        '_display_name',
+        '_parameters',
+        '_uri',
+    )
+
+    def __init__(self, header):
+        result = parse_header(header)
+        self._display_name = result.display_name
+        self._parameters = result.parameters
+        self._uri = result.uri
+
+    @classmethod
+    def build(cls, *, uri, display_name=None, parameters=None):
+        '''Build a new Header from kwargs.'''
+        self = object.__new__(cls)
+        self._uri = uri
+        self._display_name = display_name
+        self._parameters = parameters if parameters else {}
+        self._validate()
+        return self
+
+    display_name = property(lambda self: self._display_name)
+    parameters = property(lambda self: self._parameters)
+    uri = property(lambda self: self._uri)
+    tag = property(lambda self: self._parameters.get('tag', None))
+
+    def with_display_name(self, display_name):
+        new = copy.deepcopy(self)
```

Comment:
We probably don't need a deep copy of the URI. Since any changes to the URI will create a new URI, sharing a reference shouldn't be a problem.